### PR TITLE
[CDAP-14572] Adds toggle plugin widget

### PR DESCRIPTION
**Pull request by tonybach**
JIRA: https://issues.cask.co/browse/CDAP-14572
UI builds: https://builds.cask.co/browse/CDAP-UDUT149
Docs builds: https://builds.cask.co/browse/CDAP-DQB442

This PR:
- Adds `toggle` widget that can be used when building pipeline plugins. This is based on the existing `ToggleSwitch` React component. The user can specify the 'On'/'Off' labels, and the value to send to the backend for each state. Currently we set a limit of 64 characters for the labels.
- Modifies Configure modeless in Studio to use `toggle-switch` directive based on React component, instead of legacy `my-toggle-switch-widget` Angular directive.

### CODE REVIEW

File: cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst [hunk 1]
```diff
@@ -687,13 +687,44 @@ CDAP pipelines as of version |version|.
...
+     - .. container:: copyable copyable-text
+
+         ::
+
+          {
+            "name": "property-toggle-siwtch",
+            "widget-type": "toggle-switch",
+            "widget-attributes": {
+              "on": {
+                "id": "on-value",
```
Comment (elfenheart) [inline cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst]: why is this called `id`? The signature does not seem to make sense.. Should just be called `value`

File: cdap-ui/app/directives/widget-container/widget-toggle-switch/widget-toggle-switch.js [hunk 2]
```diff
@@ -14,15 +14,34 @@
...
-        onToggle: '&'
+        model: '=ngModel',
+        config: '=',
+        disabled: '=',
       },
-      templateUrl: 'widget-container/widget-toggle-switch/widget-toggle-switch.html'
+      templateUrl: 'widget-container/widget-toggle-switch/widget-toggle-switch.html',
+      controller: function($scope, myHelpers) {
+        const onValue = myHelpers.objectQuery($scope.config, 'widget-attributes', 'on', 'id') || 'on-value';
+        const offValue = myHelpers.objectQuery($scope.config, 'widget-attributes', 'off', 'id') || 'off-value';
```
Comment (elfenheart) [inline cdap-ui/app/directives/widget-container/widget-toggle-switch/widget-toggle-switch.js]: the default value here should just be `on` or `off`

File: cdap-ui/app/directives/widget-container/widget-toggle-switch/widget-toggle-switch.js [hunk 3]
```diff
@@ -14,15 +14,34 @@
...
+        const defaultValue = myHelpers.objectQuery($scope.config, 'widget-attributes', 'default') || 'on-value';
```
Comment (elfenheart) [inline cdap-ui/app/directives/widget-container/widget-toggle-switch/widget-toggle-switch.js]: the `defaultValue` if the user does not specify should be `onValue` variable

**Review by elfenheart — DISMISSED** [review]
👍 LGTM

File: cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst [hunk 4]
```diff
@@ -22,8 +22,8 @@ list of plugins shown here, as these other widgets are internal to CDAP and may
 supported in a future release.
 
 This document describes version |plugins-spec-version| of the plugin specification.
-Changes to the specification are described in the  
-:ref:`plugins-presentation-specification-changes` and should be checked if you are using a 
+Changes to the specification are described in the
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst]: undo this change

File: cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst [hunk 5]
```diff
@@ -46,7 +46,7 @@ The widget JSON is composed of:
 .. highlight:: json-ellipsis
 
 Each configuration group consists of a list of the :ref:`individual properties
-<plugins-presentation-property-configuration>` and the 
+<plugins-presentation-property-configuration>` and the
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst]: ditto

File: cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst [hunk 6]
```diff
@@ -195,8 +195,8 @@ Each individual property of the plugin is represented by a configuration, compos
 
 Note that all properties and property values are case-sensitive.
 
-To find the available field names, you can use the :ref:`Artifact HTTP RESTful API 
-<http-restful-api-artifact>` to :ref:`retrieve plugin details 
+To find the available field names, you can use the :ref:`Artifact HTTP RESTful API
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst]: ditto

File: cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst [hunk 7]
```diff
@@ -687,13 +687,44 @@ CDAP pipelines as of version |version|.
...
+
+          {
+            "name": "property-toggle-siwtch",
+            "widget-type": "toggle-switch",
+            "widget-attributes": {
+              "on": {
+                "value": "on",
+                "label": "On"
+              },
+              "on": {
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst]: "off"

**Review by bdmogal — CHANGES_REQUESTED** [review]
There's a lot of unintentional changes in the RST file. Can you fix that?

File: cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst [hunk 8]
```diff
@@ -687,13 +687,44 @@ CDAP pipelines as of version |version|.
             }
           }
 
+   * - ``toggle-switch``
+     - - ``on``: 'On' state of the switch, defined by
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst]: remove the extra `-`

File: cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst [hunk 9]
```diff
@@ -687,13 +687,44 @@ CDAP pipelines as of version |version|.
...
+         - ``value``: value to be set for the plugin property
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst]: this formatting (bullets) doesn't really work inside a table.

File: cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst [hunk 10]
```diff
@@ -687,13 +687,44 @@ CDAP pipelines as of version |version|.
...
+         - ``label``: a label to be rendered in UI for the 'On' state, limited to 64 characters.
+       - ``off``: 'Off' state of the switch, defined by
+         - ``value``: value to be set for the plugin property
+         - ``label``: a label to be rendered in UI for the 'Off' state, limited to 64 characters.
+       - ``default``: default value for the widget
+     - ``string``
+     - A toggle switch widget that allows toggling between 'on' and 'off' states, and supports setting a custom value
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst]: I think the second part of this statement is unnecessary. Let's just stick to "A toggle switch widget that allows toggling between 'on' and 'off' states". If we're talking about custom values here, why not custom labels. I think that level of detail is in the attributes description, so no need to repeat it here.

File: cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst [hunk 11]
```diff
@@ -687,13 +687,44 @@ CDAP pipelines as of version |version|.
             }
           }
 
+   * - ``toggle-switch``
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst]: why not just ``toggle``?

File: cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst (hunk 8 above)
Comment (tonybach) [reply cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst]: This extra `-` is to show the attributes (`on`, `off`, and `default`) as bullet points inside the table cell.

File: cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst [hunk 12]
```diff
@@ -687,6 +687,34 @@ CDAP pipelines as of version |version|.
             }
           }
 
+   * - ``toggle``
+     - - ``on``: 'On' state of the toggle, defined by ``value``: value to be set for the plugin property, and
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst]: I think multiple colons in the same sentence is an anti-pattern. Perhaps use ``-`` after ``value`` and ``label`` instead

File: cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst [hunk 13]
```diff
@@ -687,6 +687,34 @@ CDAP pipelines as of version |version|.
...
+         ``label``: a label to be rendered in UI for the 'On' state, limited to 64 characters.
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst]: I think ``for the 'On' state`` is redundant, since you already start the sentence with ``'On' state of the toggle, defined by...``

File: cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst [hunk 14]
```diff
@@ -687,6 +687,34 @@ CDAP pipelines as of version |version|.
...
+       - ``off``: 'Off' state of the toggle, defined by ``value``: value to be set for the plugin property, and
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/pipelines/developing-plugins/presentation-plugins.rst]: similar comments here.

**Review by bdmogal — COMMENTED** [review]
minor comments.

**Review by bdmogal — APPROVED** [review]
